### ISSUE TITLE
UX - Code Element

### ISSUE THREAD

**Issue by xmatthewx** [mozilla/webmaker-desktop#17]
Users can drop a Coder element on a tile for freely writing code that might utilize require.js and p5.js.

What's the UX?

Is this a discreet element? Or, does it expose everything within the tile? For example, if I drop in 3 photos and Coder, can I style and animate those elements? Or will I need to inspect and reference their IDs?

Note: this element will ~~probably~~ only be available on desktop/tablet, not mobile.

**Comment by thisandagain** [comment]
I think @xmatthewx is on top of this but just something to note is the folks from the team at Drexel have a lot of useful insights and research specific to this topic that we should include in the design process.

**Comment by secretrobotron** [comment]
require.js, like http://requirebin.com/ ?

**Comment by thisandagain** [comment]
Good catch @secretrobotron. :+1: on [requirebin](https://github.com/maxogden/requirebin) as the point of reference rather than require.js.

**Comment by xmatthewx** [comment]
A few assumptions:

Assumption: the code component can be dropped on a page with other visual elements (e.g. image, text block). Editing other elements reveals a GUI to adjust attributes. Editing the code element reveals a code editor. 

Assumption: by default, the code editor will be scoped to the code element itself. The code element can add html, js, and css, but by default, does not reveal the code of other elements. 

Assumption: The code element can be positioned on the page like any other element, or resized to fill the entire page. This assumption might be good or bad.

**Comment by xmatthewx** [comment]
This might be a little messy. Below are a few ideas that could be fun to prototype and valuable to test. 

#### Prototype idea 1: getElementByMagic()
- While the editor cannot view code beyond it's scope, it might target other elements or image assets with intelligent autocomplete or a visual UI (see [desk8](http://flukeout.com/examples/desk8/) button links)

#### Prototype idea 2: smart error interventions
- The paper _Towards a Taxonomy of Errors_ (@thomaspark et al), sets the stage for a toolbox of interventions. Can a code editor provide mentor-like guidance around the most common errors? I think yes!

#### Prototype idea 3: tinker, thrash & bricolage
- The paper _Designing a Transitional Web Editor_  (@thomaspark et al) describes the approach many new and experienced coders use to get a thing to work. Can a code editor better facilitate this as a productive behavior? Imagine an openhtml page with a block of code in 'test mode.' Users could edit 3 attempts side by side, toggle between them and preview the results. Not sure this fits within Webmaker.

#### Prototype idea 4: switch scope
- Beyond the code element, the editor could be used like dev tools inspector in Webmaker, to view and edit other elements one at a time as a component, or maybe the whole page. 

#### Prototype idea 5: edit via preview
- Generated preview can be visually manipulated to edit the code (e.g. element size, background color, position for animation frames). 

From my perspective, Ideas 1 and 2 are interesting and potentially useful for Webmaker. Idea 3 might be bad or awesome for openhtml.

**Comment by thisandagain** [comment]
Option 2 seems like where there is the most opportunity across products and certainly a lot of surface area for individual OSS contributors: parsing, grammar, content development, potential integration with MDN, etc, etc.

This is specific to JS but for Option 1 I'd recommend that we take a look at things like P5 and RequireBin to get a sense for how we might extend the coding environment into more playful spaces. Lots of opportunity here. :+1:

**Comment by thomaspark** [comment]
For Prototype idea 1, we've collected ideas for a JS event-driven architecture and interface to scaffold it in slides 6-10 of this presentation: https://docs.google.com/presentation/d/1Ga9cUklzBnQbhXjjsjIwhk3qHFVSnCbJp1qQncrg90w/

Prototype ideas 2 and 3 have potential, but might cloud Webmaker's unique tiles paradigm. Thimble is probably better suited to explore them, especially once the Bracket extension system is available in the new version.

**Comment by davidascher** [comment]
driveby:

He said Hypercard.  Someone give this man a cookie! (I yearn for those
specific days of yore).

I agree that jquery/processing/etc are inspirations from a simplicity of
notation POV, and I also don't want us to embed jquery.  It's really too
bad the DOM APIs are so bulky.

**Comment by davidascher** [comment]
Also: super interested in the deck, Thomas -- lots of juicy stuff in there.

**Comment by thisandagain** [comment]
I'd like to make sure we don't loose track of some of the potential opportunities that exist in leveraging MDNs API ... guiding user's to more in-depth documentation and potentially even more friendly error messages that are developed by the community.

**Comment by groovecoder** [comment]
Big :+1: for "Can a code editor provide mentor-like guidance around the most common errors? I think yes!"

We're working on similar projects to put MDN content into developers' workflows as early as possible. E.g., highlighting CSS syntax that breaks in some browsers, and providing a hint to fix and a link to the MDN article for it.

I'll watch this issue as well to see how it might eventually use the MDN "API" for integrated content. [Bug 980006](https://bugzilla.mozilla.org/show_bug.cgi?id=980006) is doing the same thing in the Firefox Developer Edition css inspector.

**Comment by thisandagain** [comment]
Awesome. Thanks @groovecoder ! :+1: :smile:

**Comment by xmatthewx** [comment]
Some prompts for a future design session on a code element:
https://docs.google.com/document/d/1ZH8mpDj_uZKPfDn2W4sVW0q_rGwLAzyPEWtb0pclMig/edit#

**Comment by xmatthewx** [comment]
Discussed in Whistler. Noted by @flukeout:  
- At what level do users edit code?
  - Page Level (special coding page type)?
  - Element Level
- Do we have a separation of HTML, CSS & JS or just one page with all of the data?
- Do we allow users to 'break apart' existing elements to edit them?
- How will we leverage Thimble to do this?
